### ISSUE TITLE
list comprehension turning `Literal` into `str`, but not `list()`

### ISSUE THREAD

**Issue by thejcannon** [microsoft/pyright#7586]
**Describe the bug**

```
from typing import Literal

ABC = Literal["abc"]

my_dict: dict[ABC, None] = {}

x = [v for v in my_dict]
reveal_type(x)  # Type of "x" is "list[str]"
y = list(v for v in my_dict)
reveal_type(y)    # Type of "x" is "list[Literal['abc']]"
```

`pyright 1.1.356`

**Comment by erictraut** [comment]
Thanks for the bug report.

I agree this is inconsistent behavior and therefore a bug. All of these cases should turn the `Literal` into a `str`. This is consistent with non-comprehension list forms like `["a"]` which is inferred as `list[str]`, not `list[Literal["a"]]`.

**Comment by erictraut** [comment]
This will be addressed in the next release.

**Comment by thejcannon** [comment]
Wait no, not like that 😳

**Comment by thejcannon** [comment]
What's the expected type-safe of preserving the type here when converting to a list (fwiw we use the comprehension for the `if`)

**Comment by erictraut** [comment]
If you want to override the default type inference rules for a list, you can assign the expression to a variable or parameter with a declared type. Then pyright will use [bidirectional type inference](https://microsoft.github.io/pyright/#/type-inference?id=bidirectional-type-inference-expected-types) rather than the [default inference rules](https://microsoft.github.io/pyright/#/type-inference?id=list-expressions).

```python
x: list[ABC] = [v for v in my_dict]
y: list[ABC] = list(v for v in my_dict)
```

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.357, which I just published.

**Comment by Aran-Fey** [comment]
This is not analogous to `["a"]` being inferred as `list[str]`. That `"a"` could be meant to be any number of things - `Literal["a"]`, `str`, `object`, `Iterable`, `Container`, or a million other things. But here you have been explicitly told that keys of `my_dict` are of type `ABC`.
